wpilibc/src/main/native/include/frc/LEDPattern.h

```diff
@@ -0,0 +1,325 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+#pragma once
+
+#include <functional>
+#include <span>
+#include <utility>
+
+#include <units/length.h>
+#include <units/time.h>
+#include <units/velocity.h>
+
+#include "frc/AddressableLED.h"
+#include "util/Color.h"
+#include "util/Color8Bit.h"
+
+namespace frc {
+
+/**
+ * Sets the LED at the given index to the given color.
+ */
+using LEDWriterFn = std::function<void(int, frc::Color)>;
+
+/**
+ * Accepts a data buffer (1st argument) and a callback (2nd argument) for
+ * writing data.
+ */
+using LEDPatternFn =
+    std::function<void(std::span<frc::AddressableLED::LEDData>, LEDWriterFn)>;
+
+class LEDPattern {
+ public:
+  explicit LEDPattern(const LEDPatternFn& impl);
+
+  ~LEDPattern() = default;
+
+  /**
+   * Writes the pattern to an LED buffer. Dynamic animations should be called
+   * periodically (such as with a command or with a periodic method) to refresh
+   * the buffer over time.
+   *
+   * This method is intentionally designed to use separate objects for reading
+   * and writing data. By splitting them up, we can easily modify the behavior
+   * of some base pattern to make it scroll, blink, or breathe by intercepting
+   * the data writes to transform their behavior to whatever we like.
+   */
+  void ApplyTo(std::span<frc::AddressableLED::LEDData> data,
+               LEDWriterFn writer);
+  void ApplyTo(std::span<frc::AddressableLED::LEDData> data);
```

Comment:
Should have comment blocks for both function variants, with `@param` to document the function parameters.